### PR TITLE
Refactor for different token kinds

### PR DESCRIPTION
**Pull request by GerritSe**
TODO: ~Implement claim validation for decoding~

### CODE REVIEW

**Comment by timoschilling** [comment]
let's make a call about this

**Comment by lagr** [comment]
I like the API of this approach.

File: lib/compeon/token/base.rb [hunk 1]
```diff
@@ -0,0 +1,71 @@
+# frozen_string_literal: true
+
+module Compeon
+  module Token
+    module Base
+      def self.attributes(**attributes_mapping)
```
Comment (David3141) [inline lib/compeon/token/base.rb]: I love this structure for our services in marketplace, but I'm not sure about introducing it here, in a relatively small gem. Why did you go that route? Couldn't a similar behavior (talking about the methods in `class_eval`) be achieved with inheritence?

Comment (lagr) [reply lib/compeon/token/base.rb]: Timo and I discussed this early this morning. *(in our opinion) this approach is way too complex for the amount of token classes we expect to have (and how much they will differ from each other). I would advocate for a mix of inheritance and composition (concerns) for this use case.

**Comment by timoschilling** [comment]
Some features from the old implementation are missing

File: lib/compeon/token/base.rb [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+            def initialize #{init_args}
+              @claims = {}
+              #{init_instance_vars}
+            end
+
+            def valid?
+              expires_at_valid? && attributes_valid?
+            end
+
+            def expires_at_valid?
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]: can be done by JWT itself

**Review by timoschilling — COMMENTED** [review]
I think we should define our needs for the tokens and then rethink the implementation

File: lib/compeon/token/base.rb (hunk 2 above)
Comment (GerritSe) [reply lib/compeon/token/base.rb]: 🤦‍♂ true

Comment (GerritSe) [reply lib/compeon/token/base.rb]: Wait, no. How? This is for when you create a new token that is not yet encoded. I don't want the implementer to be able to not provide an `exp` claim.

Comment (timoschilling) [reply lib/compeon/token/base.rb]: ok, but I still think we don't need this

**Comment by timoschilling** [comment]
I think we should make a call to discuss our requirements for this token.

File: lib/compeon/token/base.rb (hunk 2 above)
Comment (GerritSe) [reply lib/compeon/token/base.rb]: I think we do. We don't have tokens without an expiry time and we probably never will. I see no reason to expose ourselves to more possible errors.

File: test/compeon/token/decoder_test.rb [hunk 3]
```diff
@@ -0,0 +1,205 @@
...
+
+      def kind
+        'test'
+      end
+    end
+
+    attr_accessor :attribute
+
+    def initialize(attribute:, **claims)
+      super(**claims)
```
Comment (timoschilling) [inline test/compeon/token/decoder_test.rb]:
```suggestion
      super(claims)
```

File: lib/compeon/token/base.rb [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+      end
+
+      def encode(key:)
+        Compeon::Token::Encoder.new(
+          key: key,
+          token: self
+        ).encode
+      end
+
+      def reserved_claims
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]:
```suggestion
      def claims
```
There are no `unreserved` claims

File: lib/compeon/token/base.rb [hunk 5]
```diff
@@ -0,0 +1,65 @@
+# frozen_string_literal: true
+
+module Compeon
+  module Token
+    class Base
+      attr_accessor :aud, :exp, :iat, :iss, :sub
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]: Why not using the full names?

File: lib/compeon/token/decoder.rb [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+        @claim_verifications = claim_verifications
+        @encoded_token = encoded_token
+        @key = key
+        @token_klass = token_klass
+      end
+
+      def decode
+        raise DecodeError if decoded_token[:knd] != token_klass.kind
+
+        token_klass.new(**decoded_token_attributes)
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]:
```suggestion
        token_klass.new(decoded_token_attributes)
```
use `**` for cases like `token_klass.new(foo: :bar, **decoded_token_attributes)`, only if you pass in multiple hashes, `**` is only syntactic sugar for
```token_klass.new({foo: :bar}.merge(decoded_token_attributes))```

File: lib/compeon/token/decoder.rb [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+          algorithm: token_klass::JWT_ALGORITHM,
+          **compiled_claim_verifications
+        )[0].transform_keys(&:to_sym)
+      rescue JWT::DecodeError
+        raise DecodeError
+      end
+
+      def decoded_token_attributes
+        decoded_token.to_h do |attribute, value|
+          key = token_klass.attributes_mapping.key(attribute)
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]:
```suggestion
          alias = token_klass.attributes_mapping.key(attribute)
```
key is already an instance accessor

File: lib/compeon/token/decoder.rb [hunk 8]
```diff
@@ -0,0 +1,60 @@
+# frozen_string_literal: true
+
+require 'jwt'
+
+module Compeon
+  module Token
+    class DecodeError < StandardError; end
+
+    class Decoder
+      def initialize(claim_verifications: {}, encoded_token:, key:, token_klass:)
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb:10]:
```suggestion
      def initialize(sub: nil, iss: nil, aud: nil, exp: true, encoded_token:, key:, token_klass:)
```

File: lib/compeon/token/base.rb [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+        ).encode
+      end
+
+      def reserved_claims
+        {
+          aud: aud,
+          exp: exp,
+          iat: iat,
+          iss: iss,
+          sub: sub
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]: should we also support?
'nbf' (Not Before Time) Claim
'jti' (JWT ID) Claim

File: lib/compeon/token/base.rb [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+          Compeon::Token::Decoder.new(
+            claim_verifications: claim_verifications,
+            encoded_token: encoded_token,
+            key: key,
+            token_klass: self
+          ).decode
+        end
+      end
+
+      def initialize(**claims)
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]:
```suggestion
      def initialize(aud:, ...)
```
This has the benefit that you can't initialize with an wrong claim name aka typo

File: lib/compeon/token/access.rb [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+        def kind
+          'access'
+        end
+      end
+
+      attr_accessor :client_id, :role, :user_id
+
+      def initialize(client_id:, role:, user_id:, **claims)
+        super(**claims)
```
Comment (timoschilling) [inline lib/compeon/token/access.rb]:
```suggestion
        super(claims)
```

File: lib/compeon/token/access.rb [hunk 12]
```diff
@@ -0,0 +1,32 @@
+# frozen_string_literal: true
+
+module Compeon
+  module Token
+    class Access < Base
+      JWT_ALGORITHM = 'RS256'
+
+      class << self
```
Comment (timoschilling) [inline lib/compeon/token/access.rb:6]:
```suggestion
      class << self
        def algorithm
          'RS256'
        end

```

File: lib/compeon/token/base.rb [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+  module Token
+    class Base
+      attr_accessor :aud, :exp, :iat, :iss, :sub
+
+      class << self
+        def attributes
+          @attributes ||= attributes_mapping.keys.freeze
+        end
+
+        def token_attributes
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]: I think this method isn't used anymore

File: lib/compeon/token/decoder.rb [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+          **compiled_claim_verifications
+        )[0].transform_keys(&:to_sym)
+      rescue JWT::DecodeError
+        raise DecodeError
+      end
+
+      def decoded_token_attributes
+        decoded_token.to_h do |attribute, value|
+          key = token_klass.attributes_mapping.key(attribute)
+          [key || attribute, value]
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]:
```suggestion
          [alias || attribute, value]
```

File: lib/compeon/token/decoder.rb [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+          true,
+          algorithm: token_klass::JWT_ALGORITHM,
+          **compiled_claim_verifications
+        )[0].transform_keys(&:to_sym)
+      rescue JWT::DecodeError
+        raise DecodeError
+      end
+
+      def decoded_token_attributes
+        decoded_token.to_h do |attribute, value|
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]: We have a problem with `to_h` (I know I suggest it), it's only Ruby >= 2.6, and we currently need to support 2.5 in the lambdas.
We can use transform_keys here:
```ruby
decoded_token.transform_keys do |attribute|
  token_klass.attributes_mapping.key(attribute) || attribute
end
```

File: lib/compeon/token/decoder.rb [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        end
+      end
+
+      def compiled_claim_verifications
+        {}.tap do |verifications|
+          %i[aud iss sub].each do |claim|
+            next unless claim_verifications[claim]
+
+            verifications[claim] = claim_verifications[claim]
+            verifications["verify_#{claim}".to_sym] = true
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]:
```suggestion
            verifications[:"verify_#{claim}"] = true
```

File: lib/compeon/token/encoder.rb [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+        raise 'No key given.' if @key.nil?
+        raise 'Token is invalid.' unless @token.valid?
+      end
+
+      def encode
+        raw_token = build_raw_token
+
+        JWT.encode(
+          raw_token,
```
Comment (timoschilling) [inline lib/compeon/token/encoder.rb]:
```suggestion
          {
            **attributes,
            **token.claims,
            knd: token.class.kind
          },
```

File: lib/compeon/token/encoder.rb [hunk 18]
```diff
@@ -0,0 +1,43 @@
...
+        )
+      rescue JWT::EncodeError
+        raise EncodeError
+      end
+
+      private
+
+      attr_reader :token, :key
+
+      def build_raw_token
```
Comment (timoschilling) [inline lib/compeon/token/encoder.rb]:
```suggestion
      def attributes
        token.class.attributes_mapping.invert.tranform_value do |attribute|
          token.public_send(attribute)
        end
      end
```
or
```suggestion
      def attributes
        token.class.attributes_mapping.invert.tranform_value(&token.method(:public_send))
      end
```

File: lib/compeon/token/encoder.rb [hunk 19]
```diff
@@ -0,0 +1,43 @@
...
+      def initialize(key:, token:)
+        @token = token
+        @key = key
+
+        raise 'No key given.' if @key.nil?
+        raise 'Token is invalid.' unless @token.valid?
+      end
+
+      def encode
+        raw_token = build_raw_token
```
Comment (timoschilling) [inline lib/compeon/token/encoder.rb]:
```suggestion
```

File: lib/compeon/token/encoder.rb [hunk 20]
```diff
@@ -0,0 +1,43 @@
...
+module Compeon
+  module Token
+    class EncodeError < StandardError; end
+
+    class Encoder
+      def initialize(key:, token:)
+        @token = token
+        @key = key
+
+        raise 'No key given.' if @key.nil?
```
Comment (timoschilling) [inline lib/compeon/token/encoder.rb:14]: I think there is no need to make a type check, `JWT.encode` will fail in that case.

File: lib/compeon/token/access.rb [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+        def attributes_mapping
```
Comment (timoschilling) [inline lib/compeon/token/access.rb:7]:
```suggestion
        def attribute_mapping
```

File: lib/compeon/token/access.rb [hunk 22]
```diff
@@ -0,0 +1,32 @@
...
+          {
+            client_id: :cid,
+            role: :role,
+            user_id: :uid
+          }.freeze
+        end
+
+        def kind
```
Comment (timoschilling) [inline lib/compeon/token/access.rb:19]: I like this way!

File: lib/compeon/token/base.rb [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+          sub: sub
+        }
+      end
+
+      def valid?
+        expires_at_valid? && attributes_valid?
+      end
+
+      def expires_at_valid?
+        exp.is_a?(Numeric) && exp > Time.now.to_i
```
Comment (timoschilling) [inline lib/compeon/token/base.rb]:
```suggestion
        exp > Time.now.to_i
```
Then you automatically have a hint what's wrong with your `exp`
> ArgumentError: comparison of String with 1 failed

One other reason is that in this way `exp` can be something that implements `to_json` and `>`.

File: test/compeon/token/base_test.rb [hunk 24]
```diff
@@ -0,0 +1,131 @@
...
+    assert(true, token.respond_to?(:iat=))
+    assert(true, token.respond_to?(:iss=))
+    assert(true, token.respond_to?(:sub=))
+  end
+
+  def test_decode
+    mock = Minitest::Mock.new
+    mock.expect(:decode, 'decoded token')
+
+    decoder = lambda { |claim_verifications:, encoded_token:, key:, token_klass:|
```
Comment (timoschilling) [inline test/compeon/token/base_test.rb]:
```suggestion
    decoder = lambda do |claim_verifications:, encoded_token:, key:, token_klass:|
```

File: lib/compeon/token/access.rb (hunk 12 above)
Comment (lagr) [reply lib/compeon/token/access.rb:6]: Hm, I actually prefer both the name and using a constant.

File: lib/compeon/token/encoder.rb [hunk 25]
```diff
@@ -0,0 +1,43 @@
...
+      private
+
+      attr_reader :token, :key
+
+      def build_raw_token
+        {}.tap do |raw_token|
+          token.class.attributes_mapping.each do |attribute, token_attribute|
+            raw_token[token_attribute] = token.public_send(attribute)
+          end
+        end.merge(token.reserved_claims, knd: token.class.kind)
```
Comment (David3141) [inline lib/compeon/token/encoder.rb]: I think I'd prefer something like 
```ruby
        token.class.attributes_mapping
          .invert
          .transform_values { |attr| token.public_send(attr) }
          .merge(token.reserved_claims, knd: token.class.kind)
```

File: test/compeon/token/base_test.rb (hunk 24 above)
Comment (GerritSe) [reply test/compeon/token/base_test.rb]: I like this better as well. I just stuck to the marketplace. Changed it.

File: lib/compeon/token/base.rb [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+      end
+
+      def encode(key:)
+        Compeon::Token::Encoder.new(
+          key: key,
+          token: self
+        ).encode
+      end
+
+      def registered_claims
```
Comment (timoschilling) [inline lib/compeon/token/base.rb:50]:
```ruby
      def registered_claims
        registered_claims_mapping.invert.transform_values(&:send).compact
      end
```

File: lib/compeon/token/decoder.rb [hunk 27]
```diff
@@ -0,0 +1,67 @@
...
+        @token_klass = token_klass
+      end
+
+      def decode
+        raise DecodeError if decoded_token[:knd] != token_klass.kind
+
+        attributes = decoded_token_attributes
+        attributes.delete(:knd)
+
+        token_klass.new(**attributes)
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb]:
```suggestion
        token_klass.new(attributes)
```

File: lib/compeon/token/access.rb (hunk 21 above)
Comment (David3141) [reply lib/compeon/token/access.rb:7]: Not too important:
To be honest, I'm not a fan of `*_mapping` names, I'd rather have a name that tells me what is being mapped, e.g., `token_attribute_names` or similar. The `mapping` in the name doesn't add much value, whereas `token_attribute_names[:client_id]` tells me what I get

Comment (timoschilling) [reply lib/compeon/token/access.rb:7]: under `token_attribute_names` I would expect an array / the values of this hash.

File: lib/compeon/token/decoder.rb [hunk 28]
```diff
@@ -0,0 +1,67 @@
...
+        decoded_token.transform_keys do |attribute|
+          attributes_mapping.key(attribute) ||
+            registered_claims_mapping.key(attribute) ||
+            attribute
+        end
+      end
+
+      def compiled_claim_verifications
+        {}.tap do |verifications|
+          %i[aud iss sub].each do |claim|
```
Comment (timoschilling) [inline lib/compeon/token/decoder.rb:55]: some are missed here

File: lib/compeon/token/access.rb (hunk 21 above)
Comment (David3141) [reply lib/compeon/token/access.rb:7]: `token_attribute_name_mapping`? As I said, I'd also be fine with the current name.

File: lib/compeon/token/decoder.rb (hunk 28 above)
Comment (GerritSe) [reply lib/compeon/token/decoder.rb:55]: Which one?

**Review by timoschilling — APPROVED** [review]

**Pull request merged**